### PR TITLE
Ruby: Add post-update nodes for compound arguments

### PR DESCRIPTION
**Pull request by hvitved**
### Post-update nodes
Post-update nodes are used to enable flow out through arguments:

```rb
def taint_field x
  x.field = "taint"
end

taint_field(y)
sink(y.field)
```

There is flow from "taint" to the field `field` on post-update node for `y` in the call to `taint_field`, written `(post) y [field]`, and from `(post) y [field]` in the call to `taint_field` to `y [field]` in the call to `sink`, and finally to `y.field` in the call to `sink`. 

### Compound arguments
However, when the argument is a compound expression, such as
```rb
taint_field(if b then x else y)
sink(y.field)
```
we fail to track flow, since there is no connection between `(post) if b then x else y` and the access to `y` in `sink(y.field)`.

This PR fixes that, by also adding post-update nodes for `x` and `y` in `if b then x else y`, and let those nodes have *two* pre-update nodes:

1. the compound argument, `if b then x else y`; and
2. the underlying expressions, `x` and `y`, respectively.

This ensures that we get flow out of the call into both leafs (1), while still maintaining the invariant that the underlying expression is a pre-update node (2). Of course, compound expression can be nested, `if b then x else if c then y else z`, so we also need to handle that.

### Common in Ruby (?)

While compound arguments such as `if b then x else y` may not be that common, it is easy to write an argument that is:
```rb
taint_field(x)
# vs
taint_field (x)
```
In the former call, `x` is just a normal argument, but in the latter the argument is a parenthesized expression with just one child expression, `x`.

### CODE REVIEW

**Comment by tausbn** [comment]
> While compound arguments such as `if b then x else y` may not be that common, it is easy to write an argument that is:
> 
> ```ruby
> taint_field(x)
> # vs
> taint_field (x)
> ```
> 
> In the former call, `x` is just a normal argument, but in the latter the argument is a parenthesized expression with just one child expression, `x`.

Fascinating, and a bit horrifying.

I wonder, though, could this be handled at the level of AST desugaring?

**Review by tausbn — DISMISSED** [review]
Python changes look, well, harmless to me. :slightly_smiling_face:

**Review by MathiasVP — DISMISSED** [review]
C++ 👍. I just realized that I forgot to update `identical-files.json` when I added a couple of copies of the dataflow library into experimental last week. So if https://github.com/github/codeql/pull/10487 gets merged before this you might need to pull in main and update the identical files in C++'s experimental directory. Sorry!

**Comment by hvitved** [comment]
> I wonder, though, could this be handled at the level of AST desugaring?

Yes, we could in principle desugar `(x)` to just `x`, but since we would have to handle `(foo; x)` as well anyway, it seems a bit pointless to special case.

**Comment by tausbn** [comment]
> > I wonder, though, could this be handled at the level of AST desugaring?
> 
> Yes, we could in principle desugar `(x)` to just `x`, but since we would have to handle `(foo; x)` as well anyway, it seems a bit pointless to special case.

For this case, definitely. I was just thinking more generally if it would be less surprising for users to have `foo (x)` represented in the same way as `foo(x)`.

**Comment by hvitved** [comment]
> I was just thinking more generally if it would be less surprising for users to have `foo (x)` represented in the same way as `foo(x)`.

So, desugar `(x)` to `x` when it is the only argument of a call? I guess we could do that.

**Review by atorralba — DISMISSED** [review]
Java looks harmless too 👍

**Review by michaelnebel — DISMISSED** [review]
C# LGTM

**Comment by hvitved** [comment]
Rebased to resolve merge conflicts.

**Review by michaelnebel — DISMISSED** [review]
C# still looks good to me.

**Review by MathiasVP — DISMISSED** [review]

**Comment by hvitved** [comment]
Had to fix another semantic merge conflict.

**Review by MathiasVP — APPROVED** [review]

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -43,6 +43,32 @@ private class ExprNodeImpl extends ExprNode, NodeImpl {
   override string toStringImpl() { result = this.getExprNode().toString() }
 }
 
+/**
+ * Gets a node that may execute last in `n`, and which, when it executes last,
+ * will be the value of `n`.
+ */
+private CfgNodes::ExprCfgNode getALastEvalNode(CfgNodes::ExprCfgNode n) {
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:50]: When reading the comment explaining this predicate , I had expected it to be recursive. I see you take the transitive closure later.

**Review by aibaars — APPROVED** [review]
Looks good to me.

**Pull request merged**